In [12]:
#!/usr/bin/env python

In [13]:
"""Text Mining with re and NLTK"""

'Text Mining with re and NLTK'

# Lesson 12 Assignment

In [14]:
import warnings
import re
import nltk

In [15]:
warnings.filterwarnings('ignore')

## 1. Use the following codes to load the assignment12.txt Download assignment12.txt which contains file names. How many file names are in it? (10 points)

In [16]:
file = open("Assignment_12.txt" , 'r')
text1 = file.read()
file.close()

In [17]:
text1

'arxiv_annotate10_7_1.txt   arxiv_annotate10_7_2.txt   arxiv_annotate10_7_3.txt   arxiv_annotate1_13_1.txt   arxiv_annotate1_13_2.txt   arxiv_annotate1_13_3.txt   arxiv_annotate2_66_1.txt   arxiv_annotate2_66_2.txt   arxiv_annotate2_66_3.txt   arxiv_annotate3_80_1.txt   arxiv_annotate3_80_2.txt   arxiv_annotate3_80_3.txt   arxiv_annotate4_168_1.txt   arxiv_annotate4_168_2.txt   arxiv_annotate4_168_3.txt   arxiv_annotate5_240_1.txt   arxiv_annotate5_240_2.txt   arxiv_annotate5_240_3.txt   arxiv_annotate6_52_1.txt   arxiv_annotate6_52_2.txt   arxiv_annotate6_52_3.txt   arxiv_annotate7_268_1.txt   arxiv_annotate7_268_2.txt   arxiv_annotate7_268_3.txt   arxiv_annotate8_81_1.txt   arxiv_annotate8_81_2.txt   arxiv_annotate8_81_3.txt   arxiv_annotate9_279_1.txt   arxiv_annotate9_279_2.txt   arxiv_annotate9_279_3.txt   jdm_annotate10_210_1.txt   jdm_annotate10_210_2.txt   jdm_annotate10_210_3.txt   jdm_annotate1_103_1.txt   jdm_annotate1_103_2.txt   jdm_annotate1_103_3.txt   jdm_annotate2_107_

The text file consists of various file names. Each file name is separated by multiple white spaces. Since we know that each file name is seperated by white space, we can use `split` and `len` to easily get the number of file names present.

In [18]:
names = re.split('\s+', text1)
len(names)

90

There are 90 filenames present in the text.

## 2. Identify the pattern of the file names， and find out how many file names match the pattern. (20 points)


In [19]:
text1

'arxiv_annotate10_7_1.txt   arxiv_annotate10_7_2.txt   arxiv_annotate10_7_3.txt   arxiv_annotate1_13_1.txt   arxiv_annotate1_13_2.txt   arxiv_annotate1_13_3.txt   arxiv_annotate2_66_1.txt   arxiv_annotate2_66_2.txt   arxiv_annotate2_66_3.txt   arxiv_annotate3_80_1.txt   arxiv_annotate3_80_2.txt   arxiv_annotate3_80_3.txt   arxiv_annotate4_168_1.txt   arxiv_annotate4_168_2.txt   arxiv_annotate4_168_3.txt   arxiv_annotate5_240_1.txt   arxiv_annotate5_240_2.txt   arxiv_annotate5_240_3.txt   arxiv_annotate6_52_1.txt   arxiv_annotate6_52_2.txt   arxiv_annotate6_52_3.txt   arxiv_annotate7_268_1.txt   arxiv_annotate7_268_2.txt   arxiv_annotate7_268_3.txt   arxiv_annotate8_81_1.txt   arxiv_annotate8_81_2.txt   arxiv_annotate8_81_3.txt   arxiv_annotate9_279_1.txt   arxiv_annotate9_279_2.txt   arxiv_annotate9_279_3.txt   jdm_annotate10_210_1.txt   jdm_annotate10_210_2.txt   jdm_annotate10_210_3.txt   jdm_annotate1_103_1.txt   jdm_annotate1_103_2.txt   jdm_annotate1_103_3.txt   jdm_annotate2_107_

Let's go through each of the filenames from beginning to end and break down their structure. Firstly, we can see each file name begins with either "arxiv", "jdm", or "plos". The regular expression "[A-Z]{3,5}" will identify these substrings. Those characters are followed by "\_annotate", which makes our regular expression "[A-Z]{3,5}\_annotate". The final part of the filename consists of combinations of numbers, underscores and the .txt ending. The regular expression
"[0-9]+\_[0-9]+\_[0-9].txt" should complete the pattern. So, the final regex pattern would be "[A-Z]{3,5}\_annotate[0-9]+\_[0-9]+_[0-9]". We will use this regex pattern to match all file names that follow the pattern, then find the length of the matches.

In [20]:
pattern = re.compile('[A-Z]{3,5}_annotate[0-9]+_[0-9]+_[0-9].txt', flags=re.IGNORECASE)
matches = pattern.findall(text1)
len(matches)

84

There appears to be 84 file names that match the pattern we drafted.

## 3. Find out file names that don't match with the pattern you designed. (20 points)

In [21]:
non_matches = [name for name in names if (name not in matches)]
non_matches

['jdm_ann^otate3_120_1.txt',
 'jdm_anno&tate6_32_2.txt',
 'jdm_annotat#e8_177_2.txt',
 'plos_annotat*e1_6_2.txt',
 'plos_anno%tate5_1375_3.txt',
 'plos_annot@ate7_1233_2.txt']

The above output show the remaining 6 file names that do not match the regex pattern we created. We can notice that all 6 file names include a special character in the word annotate.

## 4. Use the following codes to read the text from arxiv_annotate1_13_1.txt. Normalize the words and find out their counts.

In [22]:
file = open('arxiv_annotate1_13_1.txt', 'r')
text2 = file.read()
file.close()

In [23]:
text2

'### abstract ###\nMISC\talthough the internet as level topology has been extensively studied over the past few years  little is known about the details of the as taxonomy\nMISC\tan as  node  can represent a wide variety of organizations  e g   large isp  or small private business  university  with vastly different network characteristics  external connectivity patterns  network growth tendencies  and other properties that we can hardly neglect while working on veracious internet representations in simulation environments\nAIMX\tin this paper  we introduce a radically new approach based on machine learning techniques to map all the ases in the internet into a natural as taxonomy\nOWNX\twe successfully classify  NUMBER   NUMBER  percent  of ases with expected accuracy of  NUMBER   NUMBER  percent \nOWNX\twe release to the community the as level topology dataset augmented with   NUMBER   the as taxonomy information and  NUMBER   the set of as attributes we used to classify ases\nOWNX\twe

First, let's begin by simply looking at the text to understand what we are working with. We can see there are many new lines and tabs present in the text. In between the newlines and tabs, there is either MISC, AIMX or OWNX. These words do not look like they provide any context to the excerpt. We will remove thes during our normalization process. We can also see there are section headings and #s used to format these headings. We will also remove these as formatting will not be necessary.

In [24]:
nltk.download('punkt')
tokens = nltk.word_tokenize(text2)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


Before we being normalizing, let's count the vocabulary of the text. We can do this by lowering the token we created above. Then, we will count the amount of unique words and words in general.

In [25]:
lower_tokens = list(map(lambda x: x.lower(), tokens))

In [26]:
print(f'Number of words: {len(lower_tokens)}\n', f'Number of unique words: {len(set(lower_tokens))}')


Number of words: 856
 Number of unique words: 333


Great! Now we know there about 856 words and over 1/3 of them are unique. Next, we will remove the words and characters mentioned above.

In [27]:
remove = ['#', 'misc', 'introduction', 'abstract', 'ownx', 'aimx']
clean_text = [x for x in lower_tokens if x not in remove]


Now that we have removed all unnecessary words and characters, we can use the word lemmatizer from the `nltk` library. This package will allow us to stem the words of their endings, but unlike the steming function, the word lemmatizer will only return words that are part of the dictionary. Let's perform the lemmatization then look at the differences in the words before and after lemmatizing.

In [ ]:
nltk.download('wordnet')
lem = nltk.WordNetLemmatizer()
lem_words = [lem.lemmatize(word) for word in clean_text]

In [29]:
print([x for x in lem_words if x not in clean_text])

['ha', 'year', 'detail', 'organization', 'characteristic', 'pattern', 'tendency', 'property', 'representation', 'environment', 'attribute', 'decade', 'ha', 'thousand', 'environment', 'identifier', 'ha', 'le', 'part', 'ha', 'challenge', 'organization', 'provider', 'provider', 'provider', 'it', 'dozen', 'thousand', 'host', 'element', 'switch', 'server', 'firewall', 'pattern', 'provider', 'customer', 'agreement', 'pattern', 'model', 'user', 'study', 'it', 'packet', 'user', 'it', 'lie', 'difference', 'characteristic', 'registry', 'difference', 'exploit', 'difference', 'class', 'property', 'statistic', 'result', 'result', 'class', 'experiment', 'result']


In [30]:
print([x for x in clean_text if x not in lem_words])

['has', 'years', 'details', 'organizations', 'characteristics', 'patterns', 'tendencies', 'properties', 'representations', 'environments', 'techniques', 'ases', 'ases', 'attributes', 'ases', 'decades', 'has', 'thousands', 'networks', 'environments', 'identifiers', 'has', 'less', 'parts', 'has', 'challenges', 'networks', 'organizations', 'providers', 'providers', 'providers', 'universities', 'companies', 'types', 'its', 'types', 'topologies', 'topologies', 'dozens', 'routers', 'thousands', 'hosts', 'elements', 'switches', 'servers', 'firewalls', 'types', 'topologies', 'ases', 'ases', 'types', 'types', 'ases', 'patterns', 'providers', 'customers', 'agreements', 'companies', 'types', 'ases', 'patterns', 'models', 'addresses', 'types', 'users', 'studies', 'its', 'packets', 'users', 'its', 'lies', 'ases', 'differences', 'characteristics', 'registries', 'differences', 'ases', 'types', 'exploits', 'differences', 'ases', 'classes', 'ases', 'properties', 'infrastructures', 'statistics', 'types'

There are a few key difference between the non-lemmatized and lemmatized words. Firstly, we can notice that plural words are stemmed to the singular version. For example, organizations become organization. The lemmatizer is also able to recognize the plural version of words that end in 'y' and are able to revert them to singular. For example, we can see words like properties and tendencies are reverted to property and tendency. Unfortunately, words that end in 's' but are not plural are also stemmed, resulting in words that may not be in the dictionary. For example, words like 'has' and 'less' return 'ha' and 'le'. Besides reverting plural to singular, the lemmatizer doesn't appear to recognize other endings like 'er', 'ing', and 'tion'.

In [31]:
dist = nltk.FreqDist(lem_words)

Finally, to get the counts of each word, we can use the `FreqDist` function from the `nltk` library. This function takes words as an input and returns a dictionary with the unique words as keys and their respective frequencies as values. Let's observe the results of this function starting with the most frequent words first.

In [38]:
results = [z for z in filter(lambda y:y[1]>=5, sorted(dict(dist).items(), key=lambda x:x[1], reverse=True))]
results

[('a', 47),
 ('the', 44),
 ('of', 34),
 ('and', 24),
 ('we', 20),
 ('in', 19),
 ('to', 18),
 ('number', 16),
 ('internet', 15),
 ('different', 12),
 ('that', 12),
 ('as', 12),
 ('topology', 11),
 ('type', 11),
 ('network', 10),
 ('with', 9),
 ('on', 9),
 ('is', 8),
 ('an', 7),
 ('for', 7),
 ('our', 7),
 ('level', 6),
 ('taxonomy', 6),
 ('other', 6),
 ('from', 6),
 ('section', 6),
 ('this', 5),
 ('evolution', 5)]

To no surprise, the seven most frequent words present are words that can be considered as 'stop-words'. Aside from stop-words, we can see words like 'number', 'internet', 'different', 'topology', 'type' and 'network' appear at least 10 times. Based on the context of these words, we can gage that the text has to do with a topic relating to technology or mathematics. Other frequent words such as 'taxonomy' and 'evolution' suggest the text definitely discusses the sciences.